### PR TITLE
Trpo implementation

### PR DESCRIPTION
**Pull request by plutasnyy**

### CODE REVIEW

File: pl_bolts/models/rl/trpo_model.py [hunk 1]
```diff
@@ -0,0 +1,457 @@
...
+def cli_main():
+    parent_parser = argparse.ArgumentParser(add_help=False)
+    parent_parser = Trainer.add_argparse_args(parent_parser)
+
+    parser = TRPO.add_model_specific_args(parent_parser)
+    args = parser.parse_args()
+    model = TRPO(**vars(args))
+
+    seed_everything(0)
+    # args.logger = WandbLogger(name='costamcostam', project='trpo', entity='ghostteam')
```
Comment (NaIwo) [inline pl_bolts/models/rl/trpo_model.py:451]: name xD 
Chyba coś tu trzeba zmienić lub odkomentować

File: pl_bolts/models/rl/trpo_model.py [hunk 2]
```diff
@@ -0,0 +1,457 @@
...
+            terminal = episode_step == self.max_episode_len
+            epoch_end = epoch_step == (self.batch_size - 1)
+
+            if done or terminal or epoch_end:
+                if not epoch_end:
+                    batch_data["episode_lengths"].append(episode_step)
+
+                cumulated_reward = sum(episode_rewards)
+                batch_data["returns"].append(cumulated_reward)
+                self.last_100_returns.append(cumulated_reward)
```
Comment (NaIwo) [inline pl_bolts/models/rl/trpo_model.py:206]: W sumie czemu tylko tutaj uaktualniamy ten last_100_returns (chodzi mi o zagnieżdżenie ifa)?

File: pl_bolts/models/rl/trpo_model.py [hunk 3]
```diff
@@ -0,0 +1,457 @@
...
+            action: Tensor representing chosen action
+            value: Tensor with value estimated by critic
+        """
+        pi, action = self.actor(state.float())
+        value = self.critic(state.float())
+        return pi, action, value
+
+    def training_step(self, batch: Tuple[Tensor, Tensor, Tensor, Tensor, Tensor], batch_idx: int) -> None:
+        """
+        Performs 10 epochs of updates for critic network and single TRPO stop for actor network
```
Comment (NaIwo) [inline pl_bolts/models/rl/trpo_model.py:250]: TRPO step (jeśli dobrze myślę xD)

**Review by NaIwo — COMMENTED** [review]
Zacnie

File: pl_bolts/models/rl/common/conjugate_gradient.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
+from typing import Callable
+
+import torch
+from torch import Tensor
+
+
+def conjugate_gradient(A: Callable, b: Tensor, delta: float = 0.0, max_iterations: int = 10) -> Tensor:
+    """
+
+    Args:
```
[the remark this answers is not in the document]
Comment (plutasnyy) [reply pl_bolts/models/rl/common/conjugate_gradient.py:10]: dzieki!

File: pl_bolts/models/rl/common/rewards.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        lam: lambda parameter
+
+    Returns:
+        list of advantages
+    """
+    rews = rewards + [last_value]
+    vals = values + [last_value]
+    # GAE
+    delta = [rews[i] + gamma * vals[i + 1] - vals[i] for i in range(len(rews) - 1)]
+    adv = discount_rewards(delta, gamma * lam)
```
[the remark this answers is not in the document]
Comment (plutasnyy) [reply pl_bolts/models/rl/common/rewards.py:47]: yep bo to gae, mozesz spojrzec we wzorki https://towardsdatascience.com/generalized-advantage-estimate-maths-and-code-b5d5bd3ce737
moze nazwa powinna byc inna ale juz zostawie

File: tests/models/rl/common/test_rewards.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    assert gt_qvals == qvals
+
+
+def test_advantages():
+    """Test calculation of advantages."""
+    rewards = [1, 1, 2, 1]
+    values = [1, 0, 0.5, 0]
+
+    expected_advantages = [4.38835898419875, 4.6659850975, 3.371595, 1.99]
+    # Explanation  deltas = rewards - 0.99 * values_(t+1) - values_t =
```
[the remark this answers is not in the document]
Comment (plutasnyy) [reply tests/models/rl/common/test_rewards.py:22]: nice catch

File: pl_bolts/models/rl/trpo_model.py (hunk 1 above)
Comment (plutasnyy) [reply pl_bolts/models/rl/trpo_model.py:451]: xD

File: pl_bolts/models/rl/trpo_model.py (hunk 2 above)
Comment (plutasnyy) [reply pl_bolts/models/rl/trpo_model.py:206]: Bo generalnie we wszystkich benchmarkach dajesz return w wynikach czyli łączny zysk, bez dyskontowania, posumowany we wszystkich stanach, dlatego dodajemy to dopiero po zakonczeniu epizodu, teraz sie zastanawiam czy jak przerywamy na koncu batcha to nie powinno nie dodawac (bo to nie mowi wtedy ile bysmy mieli gdybysmy grali do konca)
